[Bug]: Recursive type alias error fails to render in Jupyter Notebook

### Description

Run this in a Jupyter notebook

```python
from guppylang import guppy

InfiniteInts = guppy.type_alias("InfiniteInts", "tuple[int, InfiniteInts]")
InfiniteInts.check()
```

### Steps to Reproduce

_No response_

### Expected Behaviour

Expected a nice diagnostic:

```
Error: Unsupported (at file.py:106:0)
    | 
104 | 
105 | 
106 | InfiniteInts = guppy.type_alias("InfiniteInts", "tuple[int, InfiniteInts]")
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Recursive definitions are not supported
```

### Actual Behaviour

```
---------------------------------------------------------------------------
GuppyError                                Traceback (most recent call last)
    [... skipping hidden 1 frame]

Cell In[5], line 7
      5     return (0, make_infinite_ints())
----> 7 InfiniteInts.check()

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang/defs.py:120, in GuppyDefinition.check(self)
    119 """Type-check a Guppy definition."""
--> 120 return ENGINE.check_single(self.id)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:140, in pretty_errors.<locals>.pretty_errors_wrapped(*args, **kwargs)
    139 with exception_hook(hook):
--> 140     return f(*args, **kwargs)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/engine.py:388, in CompilationEngine.check_single(self, id)
    384 """Top-level function to kick of checking of a definition.
    385 
    386 This is the main driver behind `guppy.check()`.
    387 """
--> 388 self.check([id])

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:140, in pretty_errors.<locals>.pretty_errors_wrapped(*args, **kwargs)
    139 with exception_hook(hook):
--> 140     return f(*args, **kwargs)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/engine.py:412, in CompilationEngine.check(self, def_ids, reset)
    411 try:
--> 412     self.checked[def_id, entry_mono_args] = self.get_checked(
    413         def_id, entry_mono_args
    414     )
    415 except RequiresMonomorphizationError as e:
    416     # `RequiresMonomorphizationError` is raised whenever we cannot proceed
    417     # checking without having the monomorphization available. In that case,
    418     # we give up and prompt the user to specify the generic arguments.

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:140, in pretty_errors.<locals>.pretty_errors_wrapped(*args, **kwargs)
    139 with exception_hook(hook):
--> 140     return f(*args, **kwargs)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/engine.py:293, in CompilationEngine.get_checked(self, id, mono_args)
    292 if isinstance(defn, CheckableDef):
--> 293     defn = defn.check(Globals(DEF_STORE.frames[defn.id]))
    294 elif isinstance(defn, CheckableGenericDef):

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/definition/alias.py:87, in ParsedTypeAliasDef.check(self, globals)
     86 ctx = TypeParsingCtx(globals, allow_free_vars=True)
---> 87 check_not_recursive(self, ctx)
     88 ty = type_from_ast(self.type_ast, ctx)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/definition/util.py:130, in check_not_recursive(defn, ctx)
    129 """Raises a user error if a struct, enum, or type alias depends on itself."""
--> 130 _check_not_recursive(defn, ctx, [defn.id], set())

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/definition/util.py:141, in _check_not_recursive(defn, ctx, path, checked)
    140 if dependency.id in path:
--> 141     raise GuppyError(UnsupportedError(loc, "Recursive definitions"))
    142 if dependency.id not in checked:

<class 'str'>: (<class 'AssertionError'>, AssertionError())

During handling of the above exception, another exception occurred:

AssertionError                            Traceback (most recent call last)
File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:85, in exception_hook.<locals>.ipython_excepthook(shell, etype, value, tb, tb_offset)
     78 def ipython_excepthook(
     79     shell: Any,
     80     etype: type[BaseException],
   (...)     83     tb_offset: Any = None,
     84 ) -> Any:
---> 85     return hook(etype, value, tb)

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:131, in pretty_errors.<locals>.hook(excty, err, traceback)
    129 """Custom `excepthook` that intercepts `GuppyExceptions` for pretty printing."""
    130 if isinstance(err, GuppyError):
--> 131     sys.stderr.write(str(err))
    132     return
    134 # If it's not a GuppyError, fall back to default hook

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/error.py:24, in GuppyError.__str__(self)
     21 from guppylang_internals.engine import DEF_STORE
     23 renderer = DiagnosticsRenderer(DEF_STORE.sources)
---> 24 renderer.render_diagnostic(self.error)
     25 return (
     26     "\n".join(renderer.buffer)
     27     + "\n\nGuppy compilation failed due to 1 previous error\n"
     28 )

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/diagnostic.py:246, in DiagnosticsRenderer.render_diagnostic(self, diag)
    242     self.buffer += wrap(
    243         f"{self.level_str(diag.level)}: {msg}", self.MAX_MESSAGE_LINE_LEN
    244     )
    245 else:
--> 246     span = to_span(diag.span)
    247     level = self.level_str(diag.level)
    249     children_with_span = [
    250         (child, to_span(child.span)) for child in diag.children if child.span
    251     ]

File ~/code/guppy-docs/.venv/lib/python3.13/site-packages/guppylang_internals/span.py:113, in to_span(x)
    111     return x
    112 file, line_offset = get_file(x), get_line_offset(x)
--> 113 assert file is not None
    114 assert line_offset is not None
    115 # x.lineno and line_offset both start at 1, so we have to subtract 1

AssertionError: 
The original exception:
```

### Component

_No response_

### Environment

```markdown

```

### Additional Context

_No response_

--- a/guppylang/src/guppylang/decorator.py
+++ b/guppylang/src/guppylang/decorator.py
@@ -1,6 +1,7 @@
 import ast
 import builtins
 import inspect
+import linecache
 from collections.abc import Callable
 from types import FrameType
 from typing import (
@@ -734,8 +735,13 @@ def _parse_expr_string(ty_str: str, parse_err: str, sources: SourceMap) -> ast.e
     if caller_frame := get_calling_frame():
         info = inspect.getframeinfo(caller_frame)
         if caller_module := inspect.getmodule(caller_frame):
-            sources.add_file(info.filename)
             source_lines, _ = inspect.getsourcelines(caller_module)
+        else:
+            # inspect.getmodule can fail, for example if we are running in IPython. Fall
+            # back to linecache in that case
+            source_lines = linecache.getlines(info.filename)
+        if source_lines:
+            sources.add_file(info.filename)
             source = "".join(source_lines)
             annotate_location(expr_ast, source, info.filename, 1)
             # Modify the AST so that all sub-nodes span the entire line. We

--- a/tests/integration/notebooks/misc_notebook_tests.ipynb
+++ b/tests/integration/notebooks/misc_notebook_tests.ipynb
@@ -227,6 +227,11 @@
    "execution_count": 6,
    "id": "e67f98e9",
    "metadata": {
+    "deletable": true,
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
     "tags": [
      "raises-exception"
     ]
@@ -269,11 +274,52 @@
     "\n",
     "main.compile()"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3e46004f-54db-4101-b467-c815c3e14ed6",
+   "metadata": {},
+   "source": [
+    "Test that error locations for top-level declarations are correct. See https://github.com/Quantinuum/guppylang/issues/2039"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "61d519ca-87b4-4811-85f4-5e266ec301c8",
+   "metadata": {
+    "deletable": true,
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": [
+     "raises-exception"
+    ]
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Error: Unsupported (at <In[7]>:1:0)\n",
+      "  | \n",
+      "1 | InfiniteInts = guppy.type_alias(\"InfiniteInts\", \"tuple[int, InfiniteInts]\")\n",
+      "  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Recursive definitions are not supported\n",
+      "\n",
+      "Guppy compilation failed due to 1 previous error\n"
+     ]
+    }
+   ],
+   "source": [
+    "InfiniteInts = guppy.type_alias(\"InfiniteInts\", \"tuple[int, InfiniteInts]\")\n",
+    "InfiniteInts.check()"
+   ]
   }
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "guppylang (3.13.11)",
+   "display_name": "Python 3 (ipykernel)",
    "language": "python",
    "name": "python3"
   },
@@ -287,7 +333,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.13.11"
+   "version": "3.13.13"
   }
  },
  "nbformat": 4,

